hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/TranslateToPublicAddressProvider.java

```diff
@@ -0,0 +1,137 @@
+        translateToPublicAddress = resolve(addressProvider, members);
+    }
+
+    private boolean resolve(AddressProvider addressProvider, Collection<MemberInfo> members) {
+        if (!(addressProvider instanceof DefaultAddressProvider)) {
+            return false;
+        }
+
+        // Default value of DISCOVERY_SPI_PUBLIC_IP_ENABLED is `null` intentionally.
+        // If DISCOVERY_SPI_PUBLIC_IP_ENABLED is not set to true/false, we don't know the intention of the user,
+        // we will try to decide if we should use private/public address automatically in that case.
+        String publicIpEnabledProperty = properties.getString(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED);
+        if (publicIpEnabledProperty == null) {
+            if (members.isEmpty() || memberInternalAddressAsDefinedInClientConfig(members)) {
+                return false;
+            }
+
+            return membersReachableOnlyViaPublicAddress(members);
+        }
+        return properties.getBoolean(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED);
+    }
+
+    /**
+     * Checks if any member has its internal address as configured in ClientConfig.
+     * <p>
+     * If any member has its internal/private address the same as configured in ClientConfig, then it means that the client is
+     * able to connect to members via configured address. No need to use make any address translation.
+     */
+    boolean memberInternalAddressAsDefinedInClientConfig(Collection<MemberInfo> members) {
+        List<String> addresses = config.getNetworkConfig().getAddresses();
+        return members.stream()
+                .map(MemberInfo::getAddress)
+                .anyMatch(a -> addresses.contains(a.getHost())
+                        || addresses.contains(String.format("%s:%s", a.getHost(), a.getPort())));
+    }
+
+    /**
+     * Checks if members are reachable via public addresses, but not reachable via internal addresses.
+     * <p>
+     * We check only limited number of random members to reduce the slowdown of the startup.
+     */
+    private boolean membersReachableOnlyViaPublicAddress(Collection<MemberInfo> members) {
+        List<MemberInfo> shuffledList = new ArrayList<>(members);
+        Collections.shuffle(shuffledList);
+        Iterator<MemberInfo> iter = shuffledList.iterator();
+        for (int i = 0; i < REACHABLE_CHECK_NUMBER; i++) {
+            if (!iter.hasNext()) {
+                iter = shuffledList.iterator();
+            }
+            MemberInfo member = iter.next();
+            Address publicAddress = member.getAddressMap().get(CLIENT_PUBLIC_ENDPOINT_QUALIFIER);
+            if (publicAddress == null) {
+                return false;
+            }
+            Address internalAddress = member.getAddress();
+            if (isReachable(internalAddress, REACHABLE_ADDRESS_TIMEOUT_MILLIS)) {
+                return false;
+            }
+            if (!isReachable(publicAddress, NON_REACHABLE_ADDRESS_TIMEOUT_MILLIS)) {
```

Comment:
Current implementation forces detection of unreachable internal address and reachable public address for 3 members in public client-server comms scenario. It follows what's written in the TDD, but it may be a bit too pessimistic. We could change this check to the following:
```java
            if (isReachable(publicAddress, NON_REACHABLE_ADDRESS_TIMEOUT_MILLIS)) {
                return true;
            }
        }
        return false;
```

Let's consider the success path, i.e. situation when all members are up and running and network works as expected. With this check we would return early in the public client-server comm scenario, once we detect the situation for the first member. In scenario when a member is not reachable, it shouldn't change anything. When the client and members are located in the same private network and members have configured public address (which seems to be a rare case), we're going to do only one check, just like before. So, the change should only reduce the number of attempts done in the public client-server comms scenario.

@leszko did you consider this logic? If you did what are the benefits of the current implementation?

Reply:
Hmm, I'd keep it as it is. Here's my reasoning. Let's take AWS as an example. 
1. You start 3 EC2 Instances (by default they have both private and public addresses)
2. This mechanism checks first the private address, so it should be fine. But if there is some random lag in the network, then
3. This mechanism checks the public address and decide to use the public address for the communication

The outcome is wrong. And it can happen, because of some random network congestion. So, I'd prefer to always check 3 times. The performance impact is not high in my option, because it affects only the case when you will actually use the public address for communication.

Reply:
I didn't get the scenario. On which step the network starts misbehaving?

> The performance impact is not high in my option

It's not about performance, it's about client start time. Current implementation leads to +3s in the public client-server comms scenario (success path, i.e. when the network is healthy). The proposed change will turn it into +1s.

Reply:
> I didn't get the scenario. On which step the network starts misbehaving?
My point is just that making a decision based on just one network call is risky. And the scenario that Hazelcast has both private and public address is a common scenario.

Reply:
> My point is just that making a decision based on just one network call is risky

In fact, the decision is made based on two network requests: the first one is supposed to fail, while the second should succeed. And current implementation makes 6 networks calls.

Reply:
Yes, but it's very common that both addresses (private and public) are reachable. Then, one false negative (of private address) gives us the wrong decision. That is why I think it's risky.

Reply:
I'm certainly fine with leaving the logic as is, but we just need to decide if it's ok to add 3 seconds to client startup time (smart and non-async start modes - which are the default). @sancar WDYT?

Reply:
Hard to decide really. There is one point that might make the decision easier. 

We have one more check before the time-consuming part. We are checking if configured addresses on the client-side are the same as the member's private(internal) addresses. If matches, then we avoid time-consuming connection attempts. 
So the setups without where clients can connect internal addresses will be fast by default most of the time. I should note that there could be still slow cases like client applications prefer to use a different network interface address than members even though both are in the same LAN(I am skipping this since it is not common). 

This also means that cloud use cases will have an additional slow startup when the client is far. In this case, with zero configuration, we are choosing between not-working and slow startup. I would like to prefer slow startup over not-working. Users can always set it explicitly to avoid slow startup as a workaround.

Reply:
> We have one more check before the time-consuming part. We are checking if configured addresses on the client-side are the same as the member's private(internal) addresses.

In fact, we do another check before doing the network check. I'm talking about `addressMap` check based on the data reported by particular member. If the check is successful, it means that the member was available through member-to-member comms and reported this state some time ago.

> This also means that cloud use cases will have an additional slow startup when the client is far. In this case, with zero configuration, we are choosing between not-working and slow startup. I would like to prefer slow startup over not-working.

No, we're choosing between slightly slower startup with a (probably?) less reliable detection mechanism and significantly slower startup with a more reliable mechanism.